File: remote_sync.py
"""Remote data synchronization utilities."""

from __future__ import annotations

import logging
import os
import requests


logger = logging.getLogger(__name__)


def sync_database_to_server(db_path: str, url: str) -> None:
    """Upload the SQLite database at ``db_path`` to ``url`` via HTTP POST."""
    if not os.path.exists(db_path):
        raise FileNotFoundError(db_path)
    try:
        with open(db_path, "rb") as fh:
            files = {"file": fh}
            resp = requests.post(url, files=files, timeout=30)
            resp.raise_for_status()
        logger.info("Database %s synced to %s", db_path, url)
    except Exception as exc:  # pragma: no cover - runtime errors
        logger.error("Sync failed: %s", exc)
        raise

File: webui/src/remoteSync.js
// Remote sync utilities similar to Python remote_sync module

export async function syncDatabaseToServer(dbPath, url, { timeout = 30, retries = 3, rowRange = null } = {}) {
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      const controller = new AbortController();
      const id = setTimeout(() => controller.abort(), timeout * 1000);
      const resp = await fetch(url, {
        method: 'POST',
        body: rowRange ? JSON.stringify({ dbPath, rowRange }) : dbPath,
        signal: controller.signal,
      });
      clearTimeout(id);
      if (!resp.ok) throw new Error('upload failed');
      return true;
    } catch (e) {
      if (attempt === retries) throw e;
      await new Promise(r => setTimeout(r, 1000));
    }
  }
  return false;
}
